### ISSUE TITLE
Task grader of activity 5.3 "free a linked list of student"

### ISSUE THREAD

**Issue by GodefroyClair** [remisharrock/c-programming-with-linux-MOOC-issues-tracker#265]
Hi,
If I add my own helper function "printStudent" (no 's') to print the info corresponding to ONE student and I declare its protocol in the protocol space where I am supposed to, ie after:
`/* add any other prototypes as needed */`
), I have an Error:
`Error - It appears you have modified restricted code.`
If reintegrate the lines of code of printStudent in the printStudents function (like in the correction), the error disappear. I should be allowed (like in the previous exercices)to create my own "helper" functions.

Here is my code with "printStudent":
`#include <stdio.h>
#include <stdlib.h>

struct student {
      char name[50];
      int age;
      struct student *next;
};

struct student *createStudent(char studentName[], int studentAge);
struct student *append(struct student * end, struct student * newStudptr);
void printStudents(struct student *start);
/* add any other prototypes as needed */
void freeStudents(struct student *studPtr);
void copyStr(char *source, char *target);
void printStudent(struct student *studPtr);

int main(void) {
    struct student *start, *newStudptr, *end;
    int ageP, ageR, ageM;

    scanf("%d %d %d", &ageP, &ageR, &ageM);

    start = createStudent("Petra", ageP);
    end = start;
    newStudptr = createStudent("Remi", ageR);
    end = append(end, newStudptr);
    newStudptr = createStudent("Mike", ageM);
    end = append(end, newStudptr);

    printStudents(start);
    freeStudents(start);

    return 0;
}

/* Place your function definitions here. Be sure to include the definitions for 
   createStudent(), append(), printStudents() as well as any other functions you 
   created for the previous tasks. */
 struct student *createStudent(char studentName[], int studentAge) {
    struct student *ptr;
    ptr = (struct student *) malloc(sizeof(struct student));
    copyStr(studentName, ptr->name);
    ptr->age = studentAge;
    ptr->next = NULL;
    return ptr;
}

void copyStr(char *source, char *target) {
    int i = 0;
    while (source[i]!='\0') {
        target[i] = source[i];
        i++;
    }
    target[i] = '\0';
}

struct student * append(struct student * end, struct student * newStudptr) {
    end->next = newStudptr;
    return(end->next);
}

void printStudents(struct student *start){
    struct student * ptr = start;
    while (ptr!=NULL) {
        printStudent(ptr);
        ptr = ptr->next;
    }
    printf("\n");
}

void printStudent(struct student *studPtr){
    printf("%s is %d years old.\n",  studPtr->name, studPtr->age);
}

void freeStudents(struct student *studPtr){
    while(studPtr->next != NULL) {
        struct student *nextPtr = studPtr->next;
        free(studPtr);
        studPtr = nextPtr;
    }
}`

**Comment by GodefroyClair** [comment]
Ok, it seems even more complicated than that...
I tried to reproduce the problem and then the validated solution but I can't do the later.
See submissions from "godzilla75": 
2018-10-03 00:28:46 : submission (score 100) in language c, length 1960
and
2018-10-03 00:41:04 : submission (score 0) in language c, length 2020
What's the difference?

**Comment by chrissmiller** [comment]
Hi Godefroy - the issue is not that you are using a helper function (that's fine) - it's that you modified the main function (with the freeStudents call and removing the *tmp variable), which the problem statement asks you not to do (and verifies when checking). I replaced the main statement code but left the printStudent function usage intact and it validates fine for me :)

**Comment by GodefroyClair** [comment]
Hi Chris,
Thanks for the answer. But I think there is still an issue. First of all, I have tried several times  after resetting the code and I was careful no to touch anything but the prototypes declaration and the functions definition.

Second, if you can, have looked at the base main function that I get when I ask for a new fresh base code (hitting the button with "plus" icon):

```
#include <stdio.h>
#include <stdlib.h>

struct student {
      char name[50];
      int age;
      struct student *next;
};

struct student *createStudent(char studentName[], int studentAge);
struct student *append(struct student * end, struct student * newStudptr);
void printStudents(struct student *start);
/* add any other prototypes as needed */

int main(void) {
    struct student *start, *newStudptr, *end;
    int ageP, ageR, ageM;

    scanf("%d %d %d", &ageP, &ageR, &ageM);

    start = createStudent("Petra", ageP);
    end = start;
    newStudptr = createStudent("Remi", ageR);
    end = append(end, newStudptr);
    newStudptr = createStudent("Mike", ageM);
    end = append(end, newStudptr);

    printStudents(start);
    freeStudents(start);

    return 0;
}

/* Place your function definitions here. Be sure to include the definitions for 
   createStudent(), append(), printStudents() as well as any other functions you 
   created for the previous tasks. */
```
Also, as far as I can see, there is no *tmp variable anyhow in the main function...

**Comment by GodefroyClair** [comment]
Ok I went a bit further and I made a diff between my two codes that passes (182990460398757543) and doesn't pass (336134743677387295) - see previous post.

To sum up, the problem here is related to `printStudents()` _not_ the main function:
```
62c62
< void printStudents(struct student *start) {
---
> void printStudents(struct student *start){
65c65
<         printf("%s is %d years old.\n", ptr->name, ptr->age);
---
>         printf("%s is %d years old.\n",  ptr->name, ptr->age);
```
Indeed, if I correct the small changes between my printStudents function and the one of the solution of the previous code I now pass. 
If you want reproduce the bug (at least on my machine), just delete a space in the definition of the printStudents function and you'll see the problem:
`void printStudents(struct student *start) {`
vs
`void printStudents(struct student *start){`
Thanks!
